### DISCUSSION TITLE
HNSW Index and Insertion time

### DISCUSSION THREAD

**Discussion by mim201820** [milvus-io/milvus#35565] [Q&A and General discussion]
I'm testing Milvus and comparing it to chromadb and qdrant.

For testing, I created a script that loads 9 pdfs:
First experiment: I load each pdf and chunk it then inserting to the index. It took around 15 min for milvus, and around 4 min for chromadb and I dont know why milvus is too slow here
Second experiment: I load and chunk all dfs then insert all of it. It took around 3 min for milvus, and around 3 min for chromadb

**Comment by xiaofan-luan** [comment]
if you do it in batch it should be very fast. 
you can share your test script and we can give some adivce.
We recommend to do batch and concurrent insertion

**Reply by mim201820** [reply]
using llama-index this is how I insert nodes to my index

```
documents = create_documents()
for document in tqdm(documents, desc="Embedding documents", unit="document"):
    nodes = split_inputs([document])
    vector_store.add(nodes)
```
Noting that this is the setting for vector store:

```
MilvusVectorStore(
        uri=url,
        collection_name="test",
        dim=768,
        overwrite=True,
        index_config={"index_type": "HNSW", 'efConstruction': 100, 'M': 16},
        similarity_metric="COSINE",
        batch_size=3000
    )
```
with this code, it took 15 min for Milvus but around 3 min for Chromadb

I know that it can be faster if I insert all nodes at once but I want to know why it is slow like this in one vectordb and not in the other

Also I compared the two collections for retrieving with this code:
```

service_context = ServiceContext.from_defaults(
        llm=None,
        embed_model=embed_model
    )
 index = VectorStoreIndex.from_vector_store(
        vector_store=vector_store,
        service_context=service_context
    )
retriever = index.as_retriever()
response = retriever.retrieve(query)
```

And it took 0.5 s for MIlvus and 0.1 for Chromadb do you know why Milvus is also slow in retrieving?

Thank you in advance

**Reply by xiaofan-luan** [reply]
I think this has nothing to do with the vectorDB itself, or at least this has nothing to do with HNSW.
Usually milvus insert time is less than 100ms, I think this is also highly depend on  your deployment and batch sie.

Check list here:
1. how did you deploy chroma and milvus? what is the network round trip latency? 
2. What is the batch size? say how much data you stored? Usually we recommnend to batch no more than 1MB data in one insertion
3. Chroma stored on local disk, milvus can store it in localdisk or kafka/pulsar(Distributed mode), what is your deployment.

Also, it is more recommened to concurrently insert into milvus. Under our test, each shard could be host around 20MB/s throughput
